pallets/ocw-testing/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+    }
+
+    #[pallet::storage]
+    pub(super) type OffchainWorkerTestEnabled<T> = StorageValue<_, bool, ValueQuery>;
+
+    /// Defines the block when next unsigned transaction will be accepted.
+    ///
+    /// To prevent spam of unsigned (and unpaid!) transactions on the network,
+    /// we only allow one transaction every `T::UnsignedInterval` blocks.
+    /// This storage entry defines when new transaction is going to be accepted.
+    #[pallet::storage]
+    pub(super) type NextUnsignedAt<T: Config> = StorageValue<_, BlockNumberFor<T>, ValueQuery>;
+
+    #[pallet::genesis_config]
+    pub struct GenesisConfig<T: Config> {
+        pub _phantom_data: PhantomData<T>,
+    }
+
+    #[pallet::genesis_build]
+    impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+        fn build(&self) {
+            <OffchainWorkerTestEnabled<T>>::put(&false);
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        /// Offchain worker entry point.
+        ///
+        /// By implementing `fn offchain_worker` you declare a new offchain worker.
+        /// This function will be called when the node is fully synced and a new best block is
+        /// successfully imported.
+        /// Note that it's not guaranteed for offchain workers to run on EVERY block, there might
+        /// be cases where some blocks are skipped, or for some the worker runs twice (re-orgs),
+        /// so the code should be able to handle that.
+        fn offchain_worker(block_number: BlockNumberFor<T>) {
+            log::info!("Entering off-chain worker.");
+            // The entry point of your code called by off-chain worker
+            let res = Self::send_raw_unsigned_transaction(block_number);
+            if let Err(e) = res {
+                log::error!("Error: {}", e);
+            }
+        }
+    }
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Switches on or off the offchain worker
+        ///
+        /// Only root (or specified authority account) should be able to switch
+        /// the off-chain worker on and off to avoid enabling it by default in production
+        #[pallet::call_index(0)]
+        #[pallet::weight({0})]
+        pub fn switch_offchain_worker(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+
+            OffchainWorkerTestEnabled::<T>::put(!OffchainWorkerTestEnabled::<T>::get());
+            Ok(().into())
+        }
+
+        /// Submits unsigned transaction that emits an event
+        ///
+        /// Can be triggered only by an offchain worker
+        #[pallet::call_index(1)]
+        #[pallet::weight({0})]
+        pub fn submit_event_unsigned(
+            origin: OriginFor<T>,
+            _block_number: BlockNumberFor<T>,
+        ) -> DispatchResultWithPostInfo {
+            // This ensures that the function can only be called via unsigned transaction.
+            ensure_none(origin)?;
```

Comment:
we should only accept this if offchain worker mode is enabled?

Reply:
Why ignore? A collator can decide to include this extrinsic in any block, no?

Reply:
@chexware can you add this check here as well? Just in case

```rust
        // Make sure offchain worker testing is enabled
        let is_offchain_worker_enabled = OffchainWorkerTestEnabled::<T>::get();
        if !is_offchain_worker_enabled {
            return InvalidTransaction::Call.into();
        }
```

Reply:
Added error handling inside the extrinsic in https://github.com/moondance-labs/tanssi/pull/723/commits/d9ddc08ac7a5c17427c4cd4cf1763ff60218b52e. However, this should be handled by [validate_unsigned](https://github.com/moondance-labs/tanssi/blob/53fac57f7effd184f9d74aed3c0b32a90ef03af9/pallets/ocw-testing/src/lib.rs#L149).